### PR TITLE
pessimism/: handle shard ddl watch error

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
https://github.com/pingcap/dm/issues/545

### What is changed and how it works?
Add retry for retryable watch errors.

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

**Comment by csuzhangxc** [comment]
@lichunzhu please resolve conflicts

File: dm/master/shardddl/pessimist.go [hunk 1]
```diff
@@ -67,13 +68,59 @@ func (p *Pessimist) Start(pCtx context.Context, etcdCli *clientv3.Client) error
...
+					rev1, rev2, err = p.buildLocks(etcdCli)
+					if err != nil {
+						log.L().Error("resetWorkerEv is failed, will retry later", zap.Error(err), zap.Int("retryNum", retryNum))
+					}
+					succeed = true
+				}
+				retryNum++
+			}
+		} else {
+			log.L().Error("pessimist is failed and will quit now", zap.Error(err))
```
Comment (csuzhangxc) [inline dm/master/shardddl/pessimist.go:110]: If `err == nil`, we still `log.L().Error` here?

File: dm/master/shardddl/pessimist.go [hunk 2]
```diff
@@ -67,13 +68,59 @@ func (p *Pessimist) Start(pCtx context.Context, etcdCli *clientv3.Client) error
...
+			for !succeed {
+				select {
+				case <-ctx.Done():
+					return nil
+				case <-time.After(500 * time.Millisecond):
+					rev1, rev2, err = p.buildLocks(etcdCli)
+					if err != nil {
+						log.L().Error("resetWorkerEv is failed, will retry later", zap.Error(err), zap.Int("retryNum", retryNum))
+					}
+					succeed = true
```
Comment (csuzhangxc) [inline dm/master/shardddl/pessimist.go:105]: if `err != nil`, do we still set `succeed = true`?

Comment (lichunzhu) [reply dm/master/shardddl/pessimist.go]: revised in https://github.com/pingcap/dm/pull/555/commits/0df853328942e5cc51268c028a24d7a2f2b3c499

File: dm/master/shardddl/pessimist.go (hunk 1 above)
Comment (lichunzhu) [reply dm/master/shardddl/pessimist.go]: revised in https://github.com/pingcap/dm/pull/555/commits/0df853328942e5cc51268c028a24d7a2f2b3c499

File: dm/master/shardddl/pessimist.go [hunk 3]
```diff
@@ -67,14 +68,65 @@ func (p *Pessimist) Start(pCtx context.Context, etcdCli *clientv3.Client) error
...
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+		// TODO: handle fatal error from run
+		p.run(ctx, etcdCli, rev1, rev2)
+	}()
+
+	p.closed = false // started now.
+	p.cancel = cancel
+	p.cli = etcdCli
```
Comment (csuzhangxc) [inline dm/master/shardddl/pessimist.go:85]: I think you should update `p.cli = etcdCli` both for L85 and L122.

Comment (lichunzhu) [reply dm/master/shardddl/pessimist.go]: updated in https://github.com/pingcap/dm/pull/555/commits/63fa07b3117e3fe07b7344108ec55708cb168c73

**Pull request merged**